atc/exec/task_step.go

```diff
@@ -188,135 +184,69 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
-
-	chosenWorker, err := step.workerPool.FindOrChooseWorkerForContainer(
-		ctx,
-		logger,
-		owner,
-		containerSpec,
-		workerSpec,
-		step.strategy,
-	)
-	if err != nil {
-		return err
-	}
-
-	container, err := chosenWorker.FindOrCreateContainer(
-		ctx,
-		logger,
-		step.delegate,
-		owner,
-		step.containerMetadata,
-		containerSpec,
-		resourceTypes,
-	)
-	if err != nil {
-		return err
+	processSpec := worker.TaskProcessSpec{
+		Path:         config.Run.Path,
+		Args:         config.Run.Args,
+		Dir:          config.Run.Dir,
+		StdoutWriter: step.delegate.Stdout(),
+		StderrWriter: step.delegate.Stderr(),
 	}
 
-	exitStatusProp, err := container.Property(taskExitStatusPropertyName)
-	if err == nil {
-		logger.Info("already-exited", lager.Data{"status": exitStatusProp})
-
-		status, err := strconv.Atoi(exitStatusProp)
-		if err != nil {
-			return err
-		}
-
-		step.succeeded = (status == 0)
-
-		err = step.registerOutputs(logger, repository, config, container, step.containerMetadata)
-		if err != nil {
-			return err
-		}
-
-		return nil
-	}
-
-	processIO := garden.ProcessIO{
-		Stdout: step.delegate.Stdout(),
-		Stderr: step.delegate.Stderr(),
+	imageSpec := worker.ImageFetcherSpec{
+		ResourceTypes: resourceTypes,
+		Delegate:      step.delegate,
 	}
+	owner := db.NewBuildStepContainerOwner(step.metadata.BuildID, step.planID, step.metadata.TeamID)
 
-	process, err := container.Attach(taskProcessID, processIO)
-	if err == nil {
-		logger.Info("already-running")
-	} else {
-		logger.Info("spawning")
-
-		step.delegate.Starting(logger, config)
-
-		process, err = container.Run(
-			garden.ProcessSpec{
-				ID: taskProcessID,
-
-				Path: config.Run.Path,
-				Args: config.Run.Args,
-
-				Dir: path.Join(step.containerMetadata.WorkingDirectory, config.Run.Dir),
-
-				// Guardian sets the default TTY window size to width: 80, height: 24,
-				// which creates ANSI control sequences that do not work with other window sizes
-				TTY: &garden.TTYSpec{
-					WindowSize: &garden.WindowSize{Columns: 500, Rows: 500},
-				},
-			},
-			processIO,
+	results := make(chan worker.TaskResult)
+	events := make(chan string, 1)
+
+	go func(results chan worker.TaskResult) {
+		results <- step.workerClient.RunTaskStep(
+			ctx,
+			logger,
+			owner,
+			containerSpec,
+			workerSpec,
+			step.strategy,
+			step.containerMetadata,
+			imageSpec,
+			processSpec,
+			events,
 		)
-	}
-	if err != nil {
-		return err
-	}
-
-	logger.Info("attached")
-
-	exited := make(chan struct{})
-	var processStatus int
-	var processErr error
-
-	go func() {
-		processStatus, processErr = process.Wait()
-		close(exited)
-	}()
-
-	select {
-	case <-ctx.Done():
-		err = step.registerOutputs(logger, repository, config, container, step.containerMetadata)
-		if err != nil {
-			return err
-		}
-
-		err = container.Stop(false)
-		if err != nil {
-			logger.Error("stopping-container", err)
-		}
+	}(results)
+
+	for {
+		select {
+		case <-ctx.Done():
+			result := <-results
```

Comment:
Do we intend to block on this line ? Is there a chance where this might not return ever ?

Reply:
We did intend to block on it, but after some refactoring we were able to run `RunTaskStep` in the same goroutine so no channel blocking required